### PR TITLE
Handle all net.Conn connections correctly

### PR DESCRIPTION
**Pull request by Pepijnvi**
This change allows alternative net.Conn implementations such as in memory connections based on net.Pipe for implementation testing.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -299,7 +301,9 @@ func (co *Conn) WriteMsg(m *Msg) (err error) {
 // Write implements the net.Conn Write method.
 func (co *Conn) Write(p []byte) (n int, err error) {
 	switch co.Conn.(type) {
-	case *net.TCPConn, *tls.Conn:
+	case *net.UDPConn:
+		break
+	case net.Conn:
 		if len(p) > MaxMsgSize {
 			return 0, &Error{err: "message too large"}
```
Comment (miekg) [inline client.go]: actually this length check makes sense regardless of transport used.

File: client.go [hunk 2]
```diff
@@ -225,12 +232,7 @@ func (co *Conn) ReadMsgHeader(hdr *Header) ([]byte, error) {
 		p = make([]byte, length)
 		n, err = io.ReadFull(co.Conn, p)
 	default:
-		if co.UDPSize > MinMsgSize {
-			p = make([]byte, co.UDPSize)
-		} else {
-			p = make([]byte, MinMsgSize)
-		}
-		n, err = co.Read(p)
+		fmt.Errorf("unknown connection type: %T", co.Conn)
```
Comment (miekg) [inline client.go]: can this even happen?

File: client.go (hunk 1 above)
Comment (Pepijnvi) [reply client.go]: Agreed moved upwards.

File: client.go (hunk 2 above)
Comment (Pepijnvi) [reply client.go]: Unlikely as most connection will be catched by net.Conn anyway made that the default.

File: client.go [hunk 3]
```diff
@@ -298,12 +298,14 @@ func (co *Conn) WriteMsg(m *Msg) (err error) {
...
-	case *net.TCPConn, *tls.Conn:
-		if len(p) > MaxMsgSize {
-			return 0, &Error{err: "message too large"}
-		}
+	if len(p) > MaxMsgSize {
+		return 0, &Error{err: "message too large"}
+	}
 
+	switch co.Conn.(type) {
+	case *net.UDPConn:
```
Comment (tmthrgd) [inline client.go]: This should instead be:
```go
if u, ok := co.Conn.(*net.UDPConn); ok {
    return u.Write(p)
}

l := make([]byte, 2)
...
```

File: client.go [hunk 4]
```diff
@@ -216,21 +216,21 @@ func (co *Conn) ReadMsgHeader(hdr *Header) ([]byte, error) {
 		err error
 	)
 	switch co.Conn.(type) {
-	case *net.TCPConn, *tls.Conn:
+	case *net.UDPConn:
```
Comment (tmthrgd) [inline client.go]: I’d say replace this `switch` statement with an `if`/`else`.

Comment (tmthrgd) [reply client.go]: Switch the `*net.UDPConn` typecast with `net.PacketConn`.

**Comment by miekg** [comment]
think this handles all comments and can be merged

**Comment by miekg** [comment]
ping @Pepijnvi are you able to make these last change(s)? tx

**Pull request merged**